app/src/main/java/ar/com/wolox/android/example/ui/home/news/NewsPresenter.kt

```diff
@@ -24,38 +24,97 @@ class NewsPresenter @Inject constructor(
+                onCallFailure { view?.showNoNetworkAlert() }
             }
-            onResponseFailed { _, _ -> view?.showWrongCredentialsAlert() }
-            onCallFailure { view?.showNoNetworkAlert() }
+            view?.showNews(news)
+            activeCoroutine = false
         }
-        view?.showNews(news)
     }
 
     fun updateNews(updateInvokeMethod: UpdateInvokeMethod) = launch {
-        if (currentPage <= totalPages) {
+        if (!activeCoroutine) {
+            activeCoroutine = true
+            if (currentPage <= totalPages) {
+                networkRequest(newRepository.getNews(currentPage)) {
+                    onResponseSuccessful {
+                        totalPages = it!!.totalPages
+                        when (updateInvokeMethod) {
+                            UpdateInvokeMethod.SCROLL -> news.addAll(it.page)
+                            else -> news.addAll(0, it.page)
+                        }
+                        currentPage++
+                        if (it.page.isNotEmpty()) view?.updateNews(news) else view?.showNoNewNewsAlert()
+                    }
+                    onResponseFailed { _, _ -> view?.showWrongCredentialsAlert() }
+                    onCallFailure { view?.showNoNetworkAlert() }
+                }
+            } else {
+                view?.showTotalPagesReachedAlert()
+            }
+            view?.disableSwipeRefreshLoader()
+            activeCoroutine = false
+        }
+    }
+
+    fun sortNews() {
+        news.sortByDescending { newDate -> newDate.date }
+        view?.updateNews(news)
+    }
+
+    fun refreshNews() = launch {
+        if (!activeCoroutine) {
+            activeCoroutine = true
+            currentPage = 1
             networkRequest(newRepository.getNews(currentPage)) {
-                onResponseSuccessful {
+                onResponseSuccessful { it ->
                     totalPages = it!!.totalPages
-                    when (updateInvokeMethod) {
-                        UpdateInvokeMethod.SCROLL -> news.addAll(it.page)
-                        else -> news.addAll(0, it.page)
-                    }
+                    news = it.page
                     currentPage++
-                    if (it.page.isNotEmpty()) view?.updateNews(news) else view?.showNoNewNewsAlert()
                 }
                 onResponseFailed { _, _ -> view?.showWrongCredentialsAlert() }
                 onCallFailure { view?.showNoNetworkAlert() }
             }
-        } else {
-            view?.showTotalPagesReachedAlert()
+            view?.updateNews(news)
+            activeCoroutine = false
         }
     }
+
+    fun updateLike(newId: Int) = launch {
+        if (!activeCoroutine) {
+            activeCoroutine = true
+            networkRequest(newRepository.updateLike(newId)) {
+                onResponseSuccessful {
+                    view?.showLikeNewNotification()
+                    news.find {
```

Comment:
This is ok, but maybe you can use `news.first { ... }`. The find method can find more than one that matches the condition (but it won't happen in this app)